stan/math/prim/functor/closure_adapter.hpp

```diff
@@ -0,0 +1,190 @@
+#ifndef STAN_MATH_PRIM_FUNCTOR_CLOSURE_ADAPTER_HPP
+#define STAN_MATH_PRIM_FUNCTOR_CLOSURE_ADAPTER_HPP
+
+#include <stan/math/prim/meta/error_index.hpp>
+#include <stan/math/prim/meta/return_type.hpp>
+#include <stan/math/prim/meta/is_stan_closure.hpp>
+#include <stan/math/prim/functor/apply.hpp>
+#include <ostream>
+
+namespace stan {
+namespace math {
+namespace internal {
+
+/**
+ * A closure that wraps a C++ lambda and captures values.
+ */
+template <bool Ref, typename F, typename... Ts>
+struct base_closure {
+  using captured_scalar_t__ = return_type_t<Ts...>;
+  using ValueOf__
+      = base_closure<false, F, decltype(eval(value_of(std::declval<Ts>())))...>;
+  using CopyOf__ = base_closure<false, F, Ts...>;
+  F f_;
+  std::tuple<capture_type_t<Ts, Ref>...> captures_;
+
+  explicit base_closure(const F& f, const Ts&... args)
+      : f_(f), captures_(args...) {}
+
+  template <typename... Args>
+  auto operator()(std::ostream* msgs, const Args&... args) const {
+    return apply([this, msgs, &args...](
+                     const auto&... s) { return f_(s..., args..., msgs); },
+                 captures_);
+  }
+};
+
+/**
+ * A closure that takes rng argument.
+ */
+template <bool Ref, typename F, typename... Ts>
+struct closure_rng {
+  using captured_scalar_t__ = double;
+  using ValueOf__ = closure_rng<false, F, Ts...>;
+  using CopyOf__ = closure_rng<false, F, Ts...>;
+  F f_;
+  std::tuple<capture_type_t<Ts, Ref>...> captures_;
+
+  explicit closure_rng(const F& f, const Ts&... args)
+      : f_(f), captures_(args...) {}
+
+  template <typename Rng, typename... Args>
+  auto operator()(Rng& rng, std::ostream* msgs, const Args&... args) const {
+    return apply([this, &rng, msgs, &args...](
+                     const auto&... s) { return f_(s..., args..., rng, msgs); },
+                 captures_);
+  }
+};
+
+/**
+ * A closure that can be called with `propto` template argument.
+ */
+template <bool Propto, bool Ref, typename F, typename... Ts>
+struct closure_lpdf {
+  using captured_scalar_t__ = return_type_t<Ts...>;
+  using ValueOf__ = closure_lpdf<Propto, false, F, Ts...>;
+  using CopyOf__ = closure_lpdf<Propto, false, F, Ts...>;
+  F f_;
+  std::tuple<capture_type_t<Ts, Ref>...> captures_;
+
+  explicit closure_lpdf(const F& f, const Ts&... args)
+      : f_(f), captures_(args...) {}
+
+  template <bool propto>
+  auto with_propto() {
```

Comment:
Yes, it's an `lupdf` hack. This normalization nonsense is rather confusing so here's an example:
```stan
parameters {
  real y[100];
}
model {
  function
  real higher_lpdf(real[] x, real(real[], int, int) f_lpdf) {
    real lp = 0;
    lp += reduce_sum(f_lpdf, x, 1);  // <-- A
    lp += reduce_sum(f_lupdf, x, 1); // <-- B
    return lp;
  }
  function
  real partial_lpdf(real[] x, int s, int e) {
    return std_normal_lupdf(x|);
  }
  target += higher_lpdf( y| partial_lpdf);  // <-- 1
  target += higher_lupdf(y| partial_lupdf); // <-- 2
}
```
Using `lpdf`  instead of `lupdf` anywhere makes the return value normalized.
In the example `reduce_sum` is called four times, at (1A), (1B), (2A), (2B), and only (2B) is unnormalized.

The above compiles to C++ that looks something like
```cpp
auto higher_lpdf = from_lambda([&](auto f_lpdf) {
  var lp = 0;
  lp += reduce_sum(f_lpdf.with_propto<false>(), x, 1); // <-- A
  lp += reduce_sum(f_lpdf.with_propto<true>(), x, 1);  // <-- B
  return lp;
});
auto partial_lpdf = from_lambda([]<bool propto>(auto x, int s, int e) {
  return std_normal_lpdf<propto>(x);
});
lp_accum__.add(higher_lpdf(y, partial_lpdf.with_propto<false>()); // <-- 1
lp_accum__.add(higher_lpdf(y, partial_lpdf.with_propto<true>());  // <-- 2
```
Every time the closure object is passed to a higher-order function `Propto` records if it's in its `lpdf` or `lupdf` form.